### DISCUSSION TITLE
Is there any way to add custom properties to PDF using java EE?

### DISCUSSION THREAD

**Discussion by Nimeshkottege** [community/community#146037] [Programming Help]
### Select Topic Area

Question

### Body

i have tried with these libraries : com.adobe.xmp , org.apache.pdfbox but couldn't find a way.

This is a simple key,value map , but wont work anyway.

```
	PDDocumentInformation info = new PDDocumentInformation();
	String key = "testKey";
        String value = "testValue";
        info.setCustomMetadataValue("testKey", "testValue");
```

<img width="690" alt="Screenshot 2024-11-18 at 18 08 04" src="https://github.com/user-attachments/assets/640aa8fe-bde8-40a3-a69d-87b6b3e35553">

**Comment by Nimeshkottege** [comment]
@tyagraj777 Thanks for the valuable Response.
i have found a couple of issues that `xmp.createAndAddBasicSchema()` and `xmpParser.serializeToByteArray` doesn't exist.

**Reply by tyagraj777** [reply]
if you can share detail error, might help ...

**Reply by Nimeshkottege** [reply]
Non existence of the above methods in libraries , are the error, No any further details.

**Reply by Nimeshkottege** [reply]
i got same in chatgpt 😉

**Reply by Nimeshkottege** [reply]
Human also creates limited Bots, in this forum, answers are not supposed to be expected from bots. otherwise, there no point in coming here 😉

**Comment by Nimeshkottege** [comment]
First you need to notify There's no `setAdditionalInformation` method in itext Library, this is for your reference,
https://api.itextpdf.com/iText/java/8.0.1/com/itextpdf/kernel/pdf/PdfDocument.html

Here the corrected code for you according to your reference
---------------------

```
   PdfReader reader = new PdfReader(input);
        PdfWriter writer = new PdfWriter(new FileOutputStream(output));
        PdfDocument pdfDoc = new PdfDocument(reader, writer);

        try {
            XMPMeta xmpMeta = XMPMetaFactory.create();

            xmpMeta.setProperty("http://ns.adobe.com/pdfx/1.3/", "CustomProperty1", "CustomValue1", new PropertyOptions());
            xmpMeta.setProperty("http://ns.adobe.com/pdfx/1.3/", "CustomProperty2", "CustomValue2", new PropertyOptions());

            pdfDoc.getXmpMetadata(true);
            pdfDoc.setXmpMetadata(xmpMeta);
        } catch (Exception e) {
            e.printStackTrace();
        } finally {
            pdfDoc.close();
        }

```

I have done enough rearch and development Testing Regarding to verify practicality of This feature Enhancement. And First need to understand custom tab of document properties not automatically populated from Xmp Metadata. So I'm seeking More Relible answers from the people who working with  **Adobe developing** experts. And Our procedure of [typefi](https://github.com/typefi) you wont be get it . Thank you for your valuable time Reserved for this.
